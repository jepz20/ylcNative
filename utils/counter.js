// @flow
type Func = () => void

type CounterParams = {
  finalValue: number,
  initialValue: number,
  duration?: number,
  framesPerSecond?: number,
  setValue: (value: number) => void,
  onStart?: Func,
  onEnd?: Func
}

type Counter = {
  onCancelAnimation: Func => void
}

export function calculateSkip (
  finalValue: number,
  initialValue: number,
  duration: number = 1,
  framesPerSecond: number = 60
) {
  const delta = Math.abs(finalValue - initialValue)
  const totalFrames = duration * framesPerSecond
  const skip = Math.ceil(delta / totalFrames)
  return finalValue > initialValue ? skip : skip * -1
}

export default ({
  finalValue,
  initialValue,
  setValue,
  duration = 1,
  framesPerSecond = 60,
  onStart = () => {},
  onEnd = () => {}
}: CounterParams): Counter => {
  if (typeof finalValue !== 'number' || typeof initialValue !== 'number') {
    throw new Error('Invalid values')
  }

  let lastFrame
  const skip = calculateSkip(
    finalValue,
    initialValue,
    duration,
    framesPerSecond
  )

  const hasEnded = frameVal => {
    if (finalValue >= initialValue && frameVal >= finalValue) return true
    if (finalValue <= initialValue && frameVal <= finalValue) return true
    return false
  }

  const count = frameVal => {
    if (hasEnded(frameVal)) {
      setValue(finalValue)
      onEnd()
      return
    }
    setValue(frameVal)
    lastFrame = global.requestAnimationFrame(() => count(frameVal + skip))
  }

  const onCancelAnimation = cancelAnimation => {
    cancelAnimation()
    global.cancelAnimationFrame(lastFrame)
  }

  try {
    onStart()
    count(initialValue)
  } catch (error) {
    if (error instanceof RangeError || error.name === 'RangeError') {
      console.warn('too many recursive animations', initialValue, finalValue)
      global.requestAnimationFrame(() => setValue(finalValue))
      onEnd()
    } else {
      throw error
    }
  }

  return {
    onCancelAnimation
  }
}
