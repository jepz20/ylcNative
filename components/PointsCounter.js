// @flow

import * as React from 'react'
import { View, Text, StyleSheet } from 'react-native'
import counter from '../utils/counter'

type Props = {
  value: number
}

type State = {
  value: number
}

const DEFAULT_LENGTH = 4
class PointsCounter extends React.Component<Props, State> {
  state = {
    value: this.props.value
  }

  setValue = (val: number) => {
    this.setState({ value: val })
  }

  componentWillReceiveProps ({ value: finalValue }: Props) {
    const { value: initialValue }: Props = this.props
    if (finalValue === initialValue) return
    counter({
      finalValue,
      initialValue,
      setValue: this.setValue,
      duration: 1.5
    })
  }

  render () {
    const { value = '' } = this.state
    const len = value.toString().length
    const fontSize = 65 + (DEFAULT_LENGTH - len) * 15
    return (
      <View style={styles.container}>
        <Text style={[styles.text, { fontSize }]}>{value}</Text>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center'
  },
  text: {
    fontSize: 65,
    color: '#fff',
    fontWeight: 'bold'
  }
})

export { PointsCounter }
