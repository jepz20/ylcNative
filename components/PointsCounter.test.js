import React from 'react'
import 'react-native'
import renderer from 'react-test-renderer'
import { shallow } from 'enzyme'
import counter from '../utils/counter'
import { PointsCounter } from './PointsCounter'

jest.mock('../utils/counter', () => jest.fn())
describe('PointsCounter', () => {
  test('renders with no props ', () => {
    const pointsCounter = renderer.create(<PointsCounter />).toJSON()
    expect(pointsCounter).toMatchSnapshot()
  })
  test('renders with a value of 2 digits', () => {
    const pointsCounter = renderer.create(<PointsCounter value={20} />).toJSON()
    expect(pointsCounter).toMatchSnapshot()
  })

  test('renders with a value of 4 digits', () => {
    const pointsCounter = renderer.create(<PointsCounter value={8000} />).toJSON()
    expect(pointsCounter).toMatchSnapshot()
  })

  test('renders with a value of 5 digits', () => {
    const pointsCounter = renderer.create(<PointsCounter value={80000} />).toJSON()
    expect(pointsCounter).toMatchSnapshot()
  })

  test('counter animation is called if props change', () => {
    const pointsCounter = shallow(
      <PointsCounter value={8000} />
    )
    pointsCounter.setProps({ value: 8000 })
    expect(counter.mock.calls.length).toBe(0)
  })
  test('counter animation is not called if props value is the same', () => {
    const pointsCounter = shallow(
      <PointsCounter value={8000} />
    )
    pointsCounter.setProps({ value: 0 })
    expect(counter.mock.calls.length).toBe(1)
  })

  test('counter test value updates the state', () => {
    const pointsCounter = shallow(
      <PointsCounter value={8000} />
    )
    pointsCounter.instance().setValue(10)
    expect(pointsCounter.state().value).toBe(10)
  })
})
