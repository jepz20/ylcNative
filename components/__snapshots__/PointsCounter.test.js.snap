// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PointsCounter renders with a value of 2 digits 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 65,
          "fontWeight": "bold",
        },
        Object {
          "fontSize": 80,
        },
      ]
    }
  >
    20
  </Text>
</View>
`;

exports[`PointsCounter renders with a value of 4 digits 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 65,
          "fontWeight": "bold",
        },
        Object {
          "fontSize": 65,
        },
      ]
    }
  >
    8000
  </Text>
</View>
`;

exports[`PointsCounter renders with a value of 5 digits 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 65,
          "fontWeight": "bold",
        },
        Object {
          "fontSize": 50,
        },
      ]
    }
  >
    80000
  </Text>
</View>
`;

exports[`PointsCounter renders with no props  1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 65,
          "fontWeight": "bold",
        },
        Object {
          "fontSize": 65,
        },
      ]
    }
  />
</View>
`;
